README.md
clj-multihash
=============

[![Circle CI](https://img.shields.io/circleci/project/github/multiformats/clj-multihash.svg?style=flat-square)](https://circleci.com/gh/multiformats/clj-multihash)
[![codecov.io](https://img.shields.io/codecov/c/github/multiformats/clj-multihash.svg?style=flat-square&branch=develop)](https://codecov.io/github/multiformats/clj-multihash?branch=develop)
[![API codox](https://img.shields.io/badge/doc-API-blue.svg)](https://multiformats.github.io/clj-multihash/api/multihash.core.html)
[![marginalia docs](https://img.shields.io/badge/doc-marginalia-blue.png)](https://multiformats.github.io/clj-multihash/marginalia/uberdoc.html)

[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](https://webchat.freenode.net/?channels=%23ipfs)
[![](https://img.shields.io/badge/readme%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
[![](https://img.shields.io/badge/project-multiformats-blue.svg?style=flat-square)](https://github.com/multiformats/multiformats)

> Clojure implementation of the Multihash spec

A Clojure library implementing the
[multihash](https://github.com/multiformats/multihash) standard. This provides an
extensible, efficient binary representation for cryptographic hashes.

## Table of Contents

- [Install](#install)
- [Usage](#usage)
  - [Serialization](#serialization)
- [Maintainers](#maintainers)
- [Contribute](#contribute)
- [License](#license)

## Install

Library releases are published on Clojars. To use the latest version with
Leiningen, add the following dependency to your project definition:

[![Clojars Project](https://clojars.org/mvxcvi/multihash/latest-version.svg)](https://clojars.org/mvxcvi/multihash)

## Usage

The main focus of this library is the `Multihash` type, which serves as an
immutable representation of the output digest for a specific hash algorithm.
Multihashes are sortable, support metadata, and implement value-based equality
and hash-code generation.

```clojure
(require '[multihash.core :as multihash]
         '[multihash.digest :as digest])

; List the known hash functions:
=> (keys multihash/algorithm-codes)
(:sha1 :sha2-256 :sha2-512 :sha3 :blake2b :blake2s)

; Manually create a multihash value:
=> (multihash/create :sha1 "0f1e2d3c4b5a6978")
#<Multihash hash:sha1:0f1e2d3c4b5a6978>

; Or use one of the constructors for built-in Java implementations:
=> (keys digest/functions)
(:sha1 :sha2-256 :sha2-512)

=> (def mhash (digest/sha2-256 "foo bar baz"))
#'user/mhash

; Properties can be accessed using keyword lookups:
=> (:algorithm mhash)
:sha2-256
=> (:code mhash)
18
=> (:length mhash)
32 ; bytes

; :digest returns a *copy* of the digest bytes:
=> (:digest mhash)
#bin "29MYwcRiruhy9BEJpN/TBIhxoD3t0P4OdXztV9rW8tc="
```

### Serialization

One of the appeals of the multihash standard is that it can be serialized
without specifying an encoding for the resulting byte representation. This
library provides several formats for multihashes:

```clojure
; Multihashes render as URN-like strings:
=> (str mhash)
"hash:sha2-256:dbd318c1c462aee872f41109a4dfd3048871a03dedd0fe0e757ced57dad6f2d7"

; Directly encode a multihash into a byte array:
=> (multihash/encode mhash)
#bin "EiDb0xjBxGKu6HL0EQmk39MEiHGgPe3Q/g51fO1X2tby1w=="

; Read data to decode a multihash:
=> (= mhash (multihash/decode *1))
true

; Full hex encoding is supported:
=> (multihash/hex mhash)
"1220dbd318c1c462aee872f41109a4dfd3048871a03dedd0fe0e757ced57dad6f2d7"
=> (= mhash (multihash/decode *1))
true

; As is base58 (compatible with IPFS):
=> (multihash/base58 mhash)
"Qmd8kgzaFLGYtTS1zfF37qKGgYQd5yKcQMyBeSa8UkUz4W"
=> (= mhash (multihash/decode *1))
true
```

Decoding is implemented as a protocol, so it can be extended to other data
source types as needed.

## Maintainers

Captain: [@greglook](https://github.com/greglook).

## Contribute

Contributions welcome. Please check out [the issues](https://github.com/multiformats/clj-multihash/issues).

Check out our [contributing document](https://github.com/multiformats/multiformats/blob/master/contributing.md)
for more information on how we work, and about contributing in general. Please
be aware that all interactions related to multiformats are subject to the IPFS
[Code of Conduct](https://github.com/ipfs/community/blob/master/code-of-conduct.md).

Small note: If editing the README, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## License

This is free and unencumbered software released into the public domain.
See the [UNLICENSE](UNLICENSE) file for more information.


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file, which
follows the conventions of [keepachangelog.com](http://keepachangelog.com/).
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

...

## [2.0.3] - 2018-01-22

### Added
- Multihash objects implement the Java `Serializable` interface so they work
  better with Apache Spark.

## [2.0.2] - 2017-11-04

### Changed
- Update various plugins and dependencies.
- Upgrade to CircleCI 2.0.
- Publish coverage to codecov.

## [2.0.1] - 2016-07-26

### Changed
- Upgrade dependency versions.
- Switch to CircleCI for tests.

## [2.0.0] - 2016-03-23

This is a major release that migrates to a cross-compiled codebase supporting
oth Clojure and ClojureScript. The base conversion logic has moved to
[alphabase](https://github.com/greglook/alphabase), and the hash constructors in
the core namespace have moved to `multihash.digest` since they differ
significantly between languages.

### Changed
- Change most files to `cljc` to support cross-compilation.
- Change most exceptions to `ex-info`.
- Move the `functions`, `test`, `sha1`, `sha2-256`, and `sha2-512` vars from
  `multihash.core` to `multihash.digest`.

### Removed
- Drop `multihash.hex` and `multihash.base58` namespaces in favor of
  `mvxcvi/alphabase`.

## [1.1.0] - 2015-11-09

### Added
- Make multihash hex digest available directly via the `:hex-digest` keyword.

### Changed
- `hex/encode` returns an empty string for empty byte arrays instead of nil.
- `hex/decode` returns an empty byte array for empty strings instead of nil.

## [1.0.0] - 2015-10-29

### Added
- Hash functions now support Strings, ByteBuffers, and InputStreams in addition
  to byte arrays.
- Add `functions` map of keywords to supported hash functions.
- Add `test` function to validate a multihash against some content.

### Fixed
- Add type hints to a number of functions.

## [0.2.0] - 2015-09-26

### Added
- Add `multihash.base58` encoding functions.
- Refactor into `multihash.hex` encoding functions.
- Define a `print-method` for multihashes.
- Add support for reading a multihash from a byte stream.

### Changed
- Upgrade to clojure 1.7.0.
- Change `Multihash` fields to be prefixed with an underscore.
- Change `(:digest multihash)` to return a byte array.

### Removed
- `(:bytes multihash)` no longer returns a byte array.

## [0.1.2] - 2015-06-06

### Changed
- Rename `algorithms` map to `algorithm-codes`.
- Improve conditionals in validation functions.

## 0.1.1 - 2015-03-06

Initial project release.

[Unreleased]: https://github.com/greglook/clj-multihash/compare/2.0.3...HEAD
[2.0.3]: https://github.com/greglook/clj-multihash/compare/2.0.2...2.0.3
[2.0.2]: https://github.com/greglook/clj-multihash/compare/2.0.1...2.0.2
[2.0.1]: https://github.com/greglook/clj-multihash/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/greglook/clj-multihash/compare/1.1.0...2.0.0
[1.1.0]: https://github.com/greglook/clj-multihash/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/greglook/clj-multihash/compare/0.2.0...1.0.0
[0.2.0]: https://github.com/greglook/clj-multihash/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/greglook/clj-multihash/compare/0.1.1...0.1.2


test/multihash/hex_test.clj
(ns multihash.hex-test
  (:require
    [clojure.test :refer :all]
    [multihash.hex :as hex]
    [multihash.test-utils :refer :all]))


(deftest encoding-test
  (is (nil? (hex/encode (byte-array 0)))
      "empty byte array encodes as nil")
  (is (= "00" (hex/encode (byte-array 1)))
      "single zero byte encodes as two zero chars")
  (is (= "007f" (hex/encode (doto (byte-array 2)
                              (aset-byte 1 127))))))


(deftest decoding-test
  (is (bytes= (byte-array 0) (hex/decode ""))
      "empty string decodes as empty byte array")
  (is (bytes= (byte-array 1) (hex/decode "00"))
      "single zero char decodes as single zero byte"))


(deftest reflexive-encoding
  (dotimes [i 10]
    (let [data (random-bytes 30)
          encoded (hex/encode data)
          decoded (hex/decode encoded)]
      (is (bytes= data decoded)
          (str "Hex coding is reflexive for "
               (.toString (BigInteger. 1 data) 16))))))


test/multihash/core_test.clj
(ns multihash.core-test
  (:require
    [clojure.string :as str]
    [clojure.test :refer :all]
    [multihash.core :as multihash])
  (:import
    clojure.lang.ExceptionInfo
    java.io.ByteArrayInputStream
    java.nio.ByteBuffer))


(deftest app-specific-codes
  (is (nil? (multihash/app-code? 17.0)))
  (is (false? (multihash/app-code? 0x00)))
  (doseq [code (range 0x01 0x10)]
    (is (true? (multihash/app-code? code))
        (str "Algorithm code " code " is app-specific")))
  (doseq [code (range 0x10 0x100)]
    (is (false? (multihash/app-code? code))
        (str "Algorithm code " code " is not app-specific"))))


(deftest get-algorithms
  (is (nil? (multihash/get-algorithm 0)))
  (is (nil? (multihash/get-algorithm 0x100)))
  (is (nil? (multihash/get-algorithm :foo/bar)))
  (doseq [code (range 0x01 0x10)]
    (let [algorithm (multihash/get-algorithm code)]
      (is (= code (:code algorithm)))
      (is (keyword? (:name algorithm)))))
  (doseq [[algorithm code] multihash/algorithm-codes]
    (let [by-name (multihash/get-algorithm algorithm)
          by-code (multihash/get-algorithm code)]
      (is (= algorithm (:name by-name)))
      (is (= code (:code by-code)))
      (is (= by-name by-code)))))


(deftest constructor-validation
  (is (thrown? ExceptionInfo
               (multihash/create :no-such-algo "0beec7b8"))
      "Unknown algorith keyword should be rejected")
  (is (thrown? ExceptionInfo
               (multihash/create 0x2F "0beec7b8"))
      "Unknown numeric code should be rejected")
  (is (thrown? ExceptionInfo
               (multihash/create :sha1 nil))
      "Nil digest should be rejected")
  (is (thrown? ExceptionInfo
               (multihash/create :sha1 ""))
      "Empty digest should be rejected")
  (is (thrown? ExceptionInfo
               (multihash/create :sha1 "018zk80q"))
      "Malformed digest should be rejected")
  (is (thrown? ExceptionInfo
               (multihash/create :sha1 (byte-array 128)))
      "Digest length should be limited to 127")
  (is (thrown? ExceptionInfo
               (multihash/create :sha1 "012"))
      "Odd digest length should be rejected"))


(deftest value-semantics
  (let [a (multihash/create 0x11 "0beec7b8")
        b (multihash/create 0x11 "94a1be0c")
        c (multihash/create 0x12 "00a8b94e")
        c' (multihash/create 0x12 "00a8b94e")]
    (is (= a a) "Identical values are equal")
    (is (= c c') "Values with same code and digest are equal")
    (is (integer? (hash b)) "Hash code returns an integer")
    (is (= (hash c) (hash c')) "Equivalent objects return same hashcode")
    (is (= [a b c] (sort [c b a])) "Multihashes sort in code/digest order")))


(deftest multihash-rendering
  (is (= "hash:sha1:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33"
         (str (multihash/create :sha1 "0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33"))))
  (is (= "hash:sha2-256:2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
         (str (multihash/create :sha2-256 "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"))))
  (is (= "hash:sha1:ea347f3c5b8f0fd07b5bc95d0beecdda3c275da3"
         (str (multihash/create :sha1 "ea347f3c5b8f0fd07b5bc95d0beecdda3c275da3")))))


(deftest exercise-metadata
  (let [a (multihash/create :sha1 "dbc95275da8a3d0d0beeea3f0fd47f3cc7b55bc3")
        a' (vary-meta a assoc :foo :bar/baz)]
    (is (empty? (meta a)) "values start with empty metadata")
    (is (= :bar/baz (:foo (meta a'))) "metadata can be associated with value")
    (is (= a a') "metadata does not affect equality")))


(def examples
  "Test case examples."
  {"11140beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33"
   [0x11 :sha1 20 "0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33"]

   "11040beec7b8"
   [0x11 :sha1 4 "0beec7b8"]

   "12202c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
   [0x12 :sha2-256 32 "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"]

   "12042c26b46b"
   [0x12 :sha2-256 4 "2c26b46b"]

   "40040006b46b"
   [0x40 :blake2b 4 "0006b46b"]})


(deftest array-decoding-failures
  (is (thrown? ExceptionInfo
               (multihash/decode-array (byte-array 2)))
      "byte array must have at least three bytes")
  (is (thrown? ExceptionInfo
               (multihash/decode-array
                 (doto (byte-array 4)
                   (aset-byte 0 0x11)
                   (aset-byte 1 0))))
      "Encoded length must be positive")
  (is (thrown? ExceptionInfo
               (multihash/decode-array
                 (doto (byte-array 4)
                   (aset-byte 0 0x11)
                   (aset-byte 1 8))))
      "Encoded length must be within byte content"))


(defn stream-fixture
  "Constructs a stream with certain leading bytes for testing."
  ([code length]
   (stream-fixture code length length))
  ([code length actual]
   (let [buffer (byte-array actual)]
     (aset-byte buffer 0 code)
     (aset-byte buffer 1 length)
     (ByteArrayInputStream. buffer))))


(deftest stream-decoding-failures
  (is (thrown? ExceptionInfo
               (multihash/decode (stream-fixture 0x11 0 4)))
      "Stream with non-positive length is illegal.")
  (is (thrown? Exception
               (multihash/decode (stream-fixture 0x11 20 5)))
      "Stream without enough data throws exception.."))


(deftest example-coding
  (testing "Encoding is reflexive"
    (let [mhash (multihash/create 0x02 "0beec7b8")]
      (is (= mhash (multihash/decode (multihash/encode mhash))))))
  (doseq [[hex [code algorithm length digest]] examples]
    (let [mhash (multihash/create algorithm digest)]
      (is (= code (:code mhash)))
      (is (= algorithm (:algorithm mhash)))
      (is (= length (:length mhash)))
      (is (= digest (:hex-digest mhash)))
      (is (= hex (multihash/hex mhash))
          "Encoded multihashes match expected hex")
      (is (= mhash (multihash/decode hex))
          "Hex decodes into expected multihash")
      (let [b58 (multihash/base58 mhash)]
        (is (string? b58) "Multihash encodes to a base-58 string")
        (is (= mhash (multihash/decode b58))
            "Multihash round-trips through Base58 encoding"))
      (let [stream (ByteArrayInputStream. (multihash/encode mhash))]
        (is (= mhash (multihash/decode stream))
            "Multihash round-trips through InputStream")))))


test/multihash/test_utils.clj
(ns multihash.test-utils
  (:import
    java.security.SecureRandom))


(defn bytes=
  "Returns true if two byte sequences are the same length and have the same
  byte content."
  [a b]
  (and (= (count a) (count b))
       (every? true? (map = a b))))


(defn random-bytes
  "Returns a byte array between zero and `max-size` bytes long with random
  content."
  [max-size]
  (let [size (rand-int max-size)
        data (byte-array size)]
    (.nextBytes (SecureRandom.) data)
    data))


project.clj
(defproject mvxcvi/multihash "2.0.3"
  :description "Native Clojure implementation of the multihash standard."
  :url "https://github.com/multiformats/clj-multihash"
  :license {:name "Public Domain"
            :url "http://unlicense.org/"}

  :deploy-branches ["master"]
  :pedantic? :abort

  :aliases
  {"cljs-repl" ["run" "-m" "clojure.main" "cljs_repl.clj"]
   "cljs-test" ["doo" "phantom" "test" "once"]
   "coverage" ["with-profile" "+test,+coverage" "cloverage"]}

  :plugins
  [[lein-cljsbuild "1.1.7"]
   [lein-doo "0.1.8" :exclusions [org.clojure/clojurescript]]]

  :dependencies
  [[mvxcvi/alphabase "1.0.0"]]

  :cljsbuild
  {:builds {:test {:source-paths ["src" "test"]
                   :compiler {:output-dir "target/cljs/out"
                              :output-to "target/cljs/tests.js"
                   :main multihash.test-runner
                   :optimizations :whitespace}}}}

  :codox
  {:metadata {:doc/format :markdown}
   :source-uri "https://github.com/multiformats/clj-multihash/blob/master/{filepath}#L{line}"
   :output-path "target/doc/api"}

  :whidbey
  {:tag-types {'multihash.core.Multihash {'data/hash 'multihash.core/base58}}}

  :profiles
  {:dev
   {:dependencies [[org.clojure/clojure "1.8.0"]
                   [org.clojure/clojurescript "1.9.946"]]}

   :coverage
   {:plugins [[lein-cloverage "1.0.10"]]
    :dependencies [[org.clojure/tools.reader "1.1.0"]]}})


src/multihash/hex.clj
(ns multihash.hex
  "Functions to encode and decode bytes as hexadecimal."
  (:require
    [clojure.string :as str]))


(defn- pad-str
  "Pads a string with leading zeroes up to the given width."
  [width value]
  (let [string (str value)]
    (if (<= width (count string))
      string
      (-> width
          (- (count string))
          (repeat "0")
          (str/join)
          (str string)))))


(defn- pad-bytes
  "Pads a byte array with leading zeroes (or trims them) to ensure it has the
  given length. Returns the same array if the length is already correct, or a
  new resized array if not."
  [width ^bytes array]
  (let [array-length (alength array)]
    (if (= width array-length)
      array
      (let [array' (byte-array width)]
        (if (< array-length width)
          (System/arraycopy array 0 array' (- width array-length) array-length)
          (System/arraycopy array (- array-length width) array' 0 width))
        array'))))


(defn encode
  "Converts a byte array into a lowercase hexadecimal string."
  ^String
  [^bytes value]
  (when value
    (if (pos? (alength value))
      (let [width (* 2 (alength value))
            hex (-> (BigInteger. 1 value)
                    (.toString 16)
                    (str/lower-case))]
        (pad-str width hex))
      "")))


(defn decode
  "Parses a hexadecimal string into a byte array. Ensures that the resulting
  array is zero-padded to match the hex string length."
  ^bytes
  [^String value]
  (when value
    (if (empty? value)
      (byte-array 0)
      (let [length (/ (count value) 2)
            int-bytes (.toByteArray (BigInteger. value 16))]
        (pad-bytes length int-bytes)))))


(defn validate
  "Checks a string to determine whether it's well-formed hexadecimal digest.
  Returns an error message if the argument is invalid."
  ^String
  [digest]
  (cond
    (not (string? digest))
      (str "Value is not a string: " (pr-str digest))

    (not (re-matches #"^[0-9a-fA-F]*$" digest))
      (str "String '" digest "' is not a valid digest: "
           "contains illegal characters")

    (< (count digest) 2)
      (str "Digest must contain at least one byte")

    (> (count digest) 254)
      (str "Digest exceeds maximum supported length of 127: " (/ (count digest) 2))

    (odd? (count digest))
      (str "String '" digest "' is not a valid digest: "
           "number of characters (" (count digest) ") is odd")

    :else nil))


(defn valid?
  "Returns true if the string is valid hexadecimal."
  [digest]
  (nil? (validate digest)))


src/multihash/base58.clj
(ns multihash.base58
  "Native Base-58 encoding implementation.")


(def alphabet "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")
(def base 58N)


(defn- bytes->chars
  "Encode a byte array into a sequence of alphabet characters."
  [^bytes data]
  (loop [n (bigint (BigInteger. 1 data))
         s (list)]
    (if (< n base)
      (conj s (get alphabet n))
      (let [r (mod n base)]
        (recur
          (/ (- n r) base)
          (conj s (get alphabet r)))))))


(defn- chars->bytes
  "Decode a sequence of characters into a byte array."
  ^bytes
  [^String data]
  (->>
    (reverse data)
    (map vector (iterate (partial * base) 1))
    (reduce
      (fn [n [b c]]
        (let [v (.indexOf alphabet (str c))]
          (when (neg? v)
            (throw (IllegalArgumentException.
                     (str "Invalid character: '" c "' is not in Base58 alphabet"))))
          (+ n (* (bigint v) b))))
      0N)
    (.toBigInteger)
    (.toByteArray)))


(defn encode
  "Converts a byte array into a Base-58 string."
  ^String
  [^bytes data]
  (apply str
    (let [zeroes (count (take-while zero? data))]
      (concat
        (repeat zeroes (first alphabet))
        (if (< zeroes (count data))
          (bytes->chars data))))))


(defn decode
  "Decodes a Base-58 string into a byte array."
  ^bytes
  [^String data]
  (let [zeroes (count (take-while #{(first alphabet)} data))]
    (if (= zeroes (count data))
      (byte-array zeroes)
      (let [byte-data (chars->bytes data)
            strip-sign? (and (> (count byte-data) 1)
                          (zero? (aget byte-data 0))
                          (neg? (aget byte-data 1)))
            result (byte-array (cond-> (+ zeroes (count byte-data))
                                 strip-sign? (dec)))]
        (System/arraycopy
          byte-data
          (if strip-sign? 1 0)
          result
          zeroes
          (- (count result) zeroes))
        result))))


src/multihash/digest.clj
(ns multihash.digest
  "Digest functions for creating new multihash constructors."
  (:refer-clojure :exclude [test])
  (:require
    [multihash.core :as multihash])
  (:import
    (java.io
      InputStream
      IOException)
    java.nio.ByteBuffer
    java.security.MessageDigest))


(def functions
  "Map of supported multihash algorithm keys to hashing functions. Each function
  should take a source of binary data as the argument and return a multihash."
  {})


(defn- digest-content
  "Constructs a cryptographic digest for a given algorithm and content. Content
  may be in the form of a raw byte array, a `ByteBuffer`, or a string. Returns
  a byte array with the digest."
  ^bytes
  [digest-name content]
  (let [algo (MessageDigest/getInstance digest-name)]
    (condp instance? content
      String
        (.update algo (.getBytes ^String content))
      (Class/forName "[B")
        (.update algo ^bytes content)
      ByteBuffer
        (.update algo ^ByteBuffer content)
      InputStream
        (let [buffer (byte-array 1024)]
          (loop []
            (let [n (.read ^InputStream content buffer 0 (count buffer))]
              (when (pos? n)
                (.update algo buffer 0 n)
                (recur)))))
      (throw (IllegalArgumentException.
               (str "Don't know how to compute digest from "
                    (class content)))))
    (.digest algo)))


(defmacro defhash
  "Defines a new convenience hashing function for the given algorithm and system
  digest name."
  [algorithm digest-name]
  (let [fn-sym (symbol (name algorithm))]
    `(do
       (defn ~fn-sym
         ~(str "Calculates the " digest-name " digest of the given byte array or "
               "buffer and returns a multihash.")
         [~'content]
         (multihash/create ~algorithm (digest-content ~digest-name ~'content)))
       (alter-var-root #'functions assoc ~algorithm ~fn-sym))))


(defhash :sha1     "SHA-1")
(defhash :sha2-256 "SHA-256")
(defhash :sha2-512 "SHA-512")


(defn test
  "Determines whether a multihash is a correct identifier for some content by
  recomputing the digest for the algorithm specified in the multihash. Returns
  nil if either argument is nil, true if the digest matches, or false if not.
  Throws an exception if the multihash specifies an unsupported algorithm."
  [mhash content]
  (when (and mhash content)
    (if-let [hash-fn (get functions (:algorithm mhash))]
      (= mhash (hash-fn content))
      (throw (ex-info
               (format "No supported hashing function for algorithm %s to validate %s"
                       (:algorithm mhash) mhash)
               {:algorithm (:algorithm mhash)})))))


src/multihash/core.clj
(ns multihash.core
  "Core multihash type definition and helper methods."
  (:refer-clojure :exclude [test])
  (:require
    [clojure.string :as str]
    [alphabase.base58 :as b58]
    [alphabase.hex :as hex])
  (:import
    (java.io
      InputStream
      IOException)
    java.nio.ByteBuffer
    java.security.MessageDigest))


;; ## Hash Function Algorithms

(def ^:const algorithm-codes
  "Map of information about the available content hashing algorithms."
  {:sha1     0x11
   :sha2-256 0x12
   :sha2-512 0x13
   :sha3     0x14
   :blake2b  0x40
   :blake2s  0x41})


(defn app-code?
  "True if the given code number is assigned to the application-specfic range.
  Returns nil if the argument is not an integer."
  [code]
  (when (integer? code)
    (< 0 code 0x10)))


(defn get-algorithm
  "Looks up an algorithm by keyword name or code number. Returns `nil` if the
  value does not map to any valid algorithm."
  [value]
  (cond
    (keyword? value)
      (when-let [code (get algorithm-codes value)]
        {:code code, :name value})

    (not (integer? value))
      nil

    (app-code? value)
      {:code value, :name (keyword (str "app-" value))}

    :else
      (some #(when (= value (val %))
               {:code value, :name (key %)})
            algorithm-codes)))



;; ## Multihash Type

;; Multihash identifiers have two properties:
;;
;; - `_code` is a numeric code for an algorithm entry in `algorithm-codes` or an
;;   application-specific algorithm code.
;; - `_digest` is a string holding the hex-encoded algorithm output.
;;
;; Digest values also support metadata.
(deftype Multihash
  [^long _code ^String _digest _meta]

  Object

  (toString [this]
    (str "hash:" (name (:name (get-algorithm _code))) \: _digest))

  (equals [this that]
    (cond
      (identical? this that) true
      (instance? Multihash that)
        (and (= _code   (._code   ^Multihash that))
             (= _digest (._digest ^Multihash that)))
      :else false))

  (hashCode [this]
    (hash-combine _code _digest))


  Comparable

  (compareTo [this that]
    (cond
      (= this that) 0
      (< _code (._code ^Multihash that)) -1
      (> _code (._code ^Multihash that)) 1
      :else (compare _digest (._digest ^Multihash that))))


  clojure.lang.ILookup

  (valAt [this k not-found]
    (case k
      :code _code
      :algorithm (:name (get-algorithm _code))
      :length (/ (count _digest) 2)
      :digest (hex/decode _digest)
      :hex-digest _digest
      not-found))

  (valAt [this k]
    (.valAt this k nil))


  clojure.lang.IObj

  (meta [_] _meta)

  (withMeta [_ meta-map]
    (Multihash. _code _digest meta-map)))


(defmethod print-method Multihash
  [v ^java.io.Writer w]
  (.write w (str v)))



;; ## Constructors

;; Remove automatic constructor function.
(ns-unmap *ns* '->Multihash)


(def functions
  "Map of supported multihash algorithm keys to hashing functions. Each function
  should take a source of binary data as the argument and return a multihash."
  {})


(defn create
  "Constructs a new Multihash identifier. Accepts either a numeric algorithm
  code or a keyword name as the first argument. The digest may either by a byte
  array or a hex string."
  [algorithm digest]
  (let [algo (get-algorithm algorithm)]
    (when-not (integer? (:code algo))
      (throw (IllegalArgumentException.
               (str "Argument " (pr-str algorithm) " does not "
                    "represent a valid hash algorithm."))))
    (let [digest (if (string? digest) digest (hex/encode digest))]
      (when-let [err (hex/validate digest)]
        (throw (IllegalArgumentException. err)))
      (Multihash. (:code algo) digest nil))))


(defn- digest-content
  "Constructs a cryptographic digest for a given algorithm and content. Content
  may be in the form of a raw byte array, a `ByteBuffer`, or a string. Returns
  a byte array with the digest."
  ^bytes
  [digest-name content]
  (let [algo (MessageDigest/getInstance digest-name)]
    (condp instance? content
      String
        (.update algo (.getBytes ^String content))
      (Class/forName "[B")
        (.update algo ^bytes content)
      ByteBuffer
        (.update algo ^ByteBuffer content)
      InputStream
        (let [buffer (byte-array 1024)]
          (loop []
            (let [n (.read ^InputStream content buffer 0 (count buffer))]
              (when (pos? n)
                (.update algo buffer 0 n)
                (recur)))))
      (throw (IllegalArgumentException.
               (str "Don't know how to compute digest from "
                    (class content)))))
    (.digest algo)))


(defmacro ^:private defhash
  "Defines a new convenience hashing function for the given algorithm and system
  digest name."
  [algorithm digest-name]
  (let [fn-sym (symbol (name algorithm))]
    `(do
       (defn ~fn-sym
         ~(str "Calculates the " digest-name " digest of the given byte array or "
               "buffer and returns a multihash.")
         [~'content]
         (create ~algorithm (digest-content ~digest-name ~'content)))
       (alter-var-root #'functions assoc ~algorithm ~fn-sym))))


(defhash :sha1     "SHA-1")
(defhash :sha2-256 "SHA-256")
(defhash :sha2-512 "SHA-512")



;; ## Encoding and Decoding

(defn encode
  "Encodes a multihash into a binary representation."
  ^bytes
  [mhash]
  (let [length (:length mhash)
        encoded (byte-array (+ length 2))]
    (aset-byte encoded 0 (byte (:code mhash)))
    (aset-byte encoded 1 (byte length))
    (System/arraycopy (:digest mhash) 0 encoded 2 length)
    encoded))


(defn hex
  "Encodes a multihash into a hexadecimal string."
  ^String
  [mhash]
  (when mhash
    (hex/encode (encode mhash))))


(defn base58
  "Encodes a multihash into a Base-58 string."
  ^String
  [mhash]
  (when mhash
    (b58/encode (encode mhash))))


(defn decode-array
  "Decodes a byte array directly into multihash. Throws `IOException` on
  malformed input, and `IllegalArgumentException` if the multihash is invalid."
  [^bytes encoded]
  (when (< (alength encoded) 3)
    (throw (IOException.
             (str "Cannot read multihash from byte array: " (alength encoded)
                  " is less than the minimum of 3"))))
  (let [code (aget encoded 0)
        length (aget encoded 1)
        payload (- (alength encoded) 2)]
    (when-not (pos? length)
      (throw (IOException.
               (str "Encoded length " length " is invalid"))))
    (when (< payload length)
      (throw (IOException.
               (str "Encoded digest length " length " exceeds actual "
                    "remaining payload of " payload " bytes"))))
    (let [digest (byte-array length)]
      (System/arraycopy encoded 2 digest 0 length)
      (create code digest))))


(defn- read-stream-digest
  "Reads a byte digest array from an input stream. First reads a byte giving
  the length of the digest data to read. Throws an IOException if the length is
  invalid or there is an error reading from the stream."
  ^bytes
  [^InputStream input]
  (let [length (.read input)]
    (when-not (pos? length)
      (throw (IOException.
               (format "Byte %02x is not a valid digest length."
                       length))))
    (let [digest (byte-array length)]
      (loop [offset 0
             remaining length]
        (let [n (.read input digest offset remaining)]
          (if (< n remaining)
            (recur (+ offset n) (- remaining n))
            digest))))))


(defprotocol Decodable
  "This protocol provides a method for data sources which a multihash can be
  read from."

  (decode
    [source]
    "Attempts to read a multihash value from the data source."))


(extend-protocol Decodable

  (class (byte-array 0))

  (decode
    [source]
    (decode-array source))


  String

  (decode
    [source]
    (decode-array
      (if (hex/valid? source)
        (hex/decode source)
        (b58/decode source))))


  InputStream

  (decode
    [source]
    (let [code (.read source)
          digest (read-stream-digest source)]
      (create code digest))))



;; ## Utility Functions

(defn test
  "Determines whether a multihash is a correct identifier for some content by
  recomputing the digest for the algorithm specified in the multihash. Returns
  nil if either argument is nil, true if the digest matches, or false if not.
  Throws an exception if the multihash specifies an unsupported algorithm."
  [mhash content]
  (when (and mhash content)
    (if-let [hash-fn (get functions (:algorithm mhash))]
      (= mhash (hash-fn content))
      (throw (RuntimeException.
               (str "No supported hashing function for algorithm "
                    (:algorithm mhash) " to validate " mhash))))))
